org.eclipse.jgit/src/org/eclipse/jgit/internal/storage/dfs/DfsBlockCaffeineCache.java

```diff
@@ -0,0 +1,244 @@
+                // plus the key/entry reference itself
+                // DfsPackKey: {
+                //   final int hash;                4 bytes
+                //   final AtomicLong cachedSize;   28 bytes
+                // }
+                // final long position;             8 bytes
+                // final int size;                  4 bytes
+                // key, value reference             8 * 2 bytes
+                .weigher((DfsPackKeyWithPosition keyWithPosition, Ref ref) -> ref == null? 60 : 60 + ref.getSize())
+                .recordStats()
+                .build();
+    }
+
+    private void cleanUpIndices(DfsPackKeyWithPosition keyWithPosition, Ref ref, RemovalCause cause) {
+        if (keyWithPosition != null && ref != null) {
+            DfsPackKey key = keyWithPosition.getDfsPackKey();
+            long position = keyWithPosition.getPosition();
+
+            if (position < 0) {
+                // if it's an index, remove the whole packFile
+                cleanUpIndicesIfExists(key);
+            } else {
+                // if it's not an index, decrease the cached size
+                // remove the whole packFile if cachedSize is below 0
+                if (key.cachedSize.addAndGet(-ref.getSize()) <= 0) {
+                    cleanUpIndicesIfExists(key);
+                }
+            }
+        }
+    }
+
+    private void cleanUpIndicesIfExists(DfsPackKey key) {
+        if (key != null) {
+            DfsPackDescription description = reversePackDescriptionIndex.remove(key);
+            if (description != null) {
+                packFileCache.remove(description);
+            }
+        }
+    }
+
+    boolean shouldCopyThroughCache(long length) {
+        return length <= maxStreamThroughCache;
+    }
+
+    DfsPackFile getOrCreate(DfsPackDescription description, DfsPackKey key) {
+        return packFileCache.compute(description, (DfsPackDescription k, DfsPackFile v) -> {
+            if (v != null && !v.invalid()) {
+                return v;
+            }
+            DfsPackKey newPackKey = key != null ? key : new DfsPackKey();
+            reversePackDescriptionIndex.put(newPackKey, description);
+            return new DfsPackFile(this, description, newPackKey);
+        });
+    }
+
+    int getBlockSize() {
+        return blockSize;
+    }
+
+    // do something when the block is invalid
+    DfsBlock getOrLoad(DfsPackFile pack, long position, DfsReader dfsReader) throws IOException {
+        final long requestedPosition = position;
+        position = pack.alignToBlock(position);
+        DfsPackKey key = pack.key;
+
+        Ref<DfsBlock> loadedBlockRef = dfsBlockAndIndicesCache.get(new DfsPackKeyWithPosition(key, position), keyWithPosition -> {
+            try {
+                DfsBlock loadedBlock = pack.readOneBlock(keyWithPosition.getPosition(), dfsReader);
+                key.cachedSize.getAndAdd(loadedBlock.size());
```

Comment:
Let's make sure that this use of "cachedSize" is documented.  I think the caching/memory strategy being laid out in a top-level class comment would be sensible.  You + I have chatted offline about stuff a bunch and it would be good to make sure that's not lost :D